### PR TITLE
Discovery Server fix reconnection [12522]

### PR DESCRIPTION
**Pull request by jparisu**
Merge after https://github.com/eProsima/Fast-DDS/pull/2244

### CODE REVIEW

**Comment by jparisu** [comment]
Test with discovery server PR https://github.com/eProsima/Discovery-Server/pull/45

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 1]
```diff
@@ -1755,25 +1755,21 @@ void DiscoveryDataBase::unmatch_participant_(
                 "Attempting to unmatch an unexisting participant: " << guid_prefix);
     }
 
-    // For each relevant participant make not relevant
-    for (eprosima::fastrtps::rtps::GuidPrefix_t relevant_participant : pit->second.relevant_participants())
+    // For each participant remove it
+    // IMPORTANT: This is not for every relevant participant, as participant A could be in other participant's B info
+    // and B not be relevant for A
+    // Using for with a map does copy the values, do not give references. Find is needed.
+    for (auto& participant_it : participants_)
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:1762]: :+1:

**Review by MiguelCompany — DISMISSED** [review]
Code looks good to me, but I still have to verify with the new tests.

**Comment by MiguelCompany** [comment]
Discovery Server CI:
[Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1372)

**Comment by jparisu** [comment]
It seems like some tests are failing due to time limits, because executing (`with SHM` and `without SHM`) one of them fails and the other don't.
Let me check in my local and change the times for a wider timeout.

**Comment by jparisu** [comment]
The time has been slightly increased, it should be enough.
Please @MiguelCompany could you launch the tests again?

**Comment by MiguelCompany** [comment]
Discovery Server CI *with new timeouts*:
[Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1376)

**Comment by MiguelCompany** [comment]
> Discovery Server CI _with new timeouts_: [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1376)

This is just failing setting a PR status, as it was not launched from a PR.

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by EduPonz** [comment]
CI with corresponding Discovery Server tests
[Build Status](http://ci.intranet.eprosima.com:8080/job/discovery_server_linux/1397/)

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 2]
```diff
@@ -1755,25 +1755,20 @@ void DiscoveryDataBase::unmatch_participant_(
                 "Attempting to unmatch an unexisting participant: " << guid_prefix);
     }
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:1756]: I think this find is not necessary

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Review by EduPonz — APPROVED** [review]
LGTM, I'll run a CI just in case

**Comment by EduPonz** [comment]
Discovery Server tests with eProsima/Discovery-Server#45:

[Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1403)

**Comment by EduPonz** [comment]
> [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1403)

I'm concerned about the `test_39_trivial_reconnect.SHM_ON` failure. What do you think @jparisu

**Comment by jparisu** [comment]
> > [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1403)
> 
> I'm concerned about the `test_39_trivial_reconnect.SHM_ON` failure. What do you think @jparisu?

I could replicate the error in local, and it is due to time limit inside the test. Server B dies before Server A Data(Up) could arrive (this could happen because they are different processes and the time could not fit correctly).
Commit https://github.com/eProsima/Discovery-Server/pull/45/commits/d72162fab6bc0e982b53bf457d3db70f3d021178 in tests extends the time of the test. This should not fail again.

**Comment by EduPonz** [comment]
Nice! One more CI run then with eProsima/Discovery-Server#45:

[Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1407)

**Comment by EduPonz** [comment]
CI run with updated eProsima/Discovery-Server#45:

[Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/1421)

**Comment by EduPonz** [comment]
@Mergifyio backport 2.1.x

**Comment by rsanchez15** [comment]
@Mergifyio backport 2.3.x

**Pull request merged**